### PR TITLE
[FrameworkBundle] add "mapping" configuration key at validation secti…

### PR DESCRIPTION
**Pull request by davewwww**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #15655 |
| License | MIT |
| Doc PR | https://github.com/symfony/symfony-docs/pull/7407 |

This feature allows you, to define additional validation files or directories which are not in the 'Bundle*/Resources/config/' directory.

``` yaml
#config.yml
framework:
  validation:
    mapping:
      paths:
        - "path/to/file/validation.yml"
        - "path/to/file/validation.xml"
        - "path/to/another/directory"
```

### CODE REVIEW

**Comment by GromNaN** [comment]
Supporting files and directories in the same list would be simpler:
```yaml
#config.yml
framework:
  validation:
    mapping:
      - "path/to/file/validation.yml"
      - "path/to/file/validation.xml"
      - "path/to/another/directory"
```

**Comment by iltar** [comment]
Check out how the translation does it, would be nice if it's got a similar configuration implementation.

**Comment by davewwww** [comment]
if you prefere to combine the files and dirs then i can merge it.
if you guys prefer to combine files and dirs, then I can do that.

**Comment by iltar** [comment]
https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L454
https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L704

At the translator part it called `path(s)`, but I think it only loads from a root dir. Why not let the validation behave the same?

**Comment by davewwww** [comment]
Is there anything I can do to finish this ticket?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -797,44 +800,82 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
...
-        return $files;
+    /**
+     * Search for validation files in directory.
+     *
+     * @param string $dir   The directory that contains the validation files
+     * @param array  $files A array reference of validation files
+     */
+    private function getValidatorMappingFilesFromDir($dir, array &$files)
+    {
+        foreach (Finder::create()->files()->in($dir)->name('/\.xml|yml$/') as $file) {
```
Comment (Ener-Getick) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: ``yaml`` should be supported I guess ?

**Comment by Ener-Getick** [comment]
@iltar the translations are meant to be put in a folder as there are at least two versions of the same file but the validation files can be in the root of a config folder so imo we should support files and dirs.

@davewwww isn't it possible to merge the two config options under ``paths`` ?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: indeed

Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Shouldn't it be `'/\.(xml|yml)$/'` ? 
(so `'/\.(xml|yml|yaml)$/'` with the `.yaml` extension)

Comment (Ener-Getick) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: You can use ``ya?ml`` too

**Comment by davewwww** [comment]
i've merged the two "files" and "dirs" options to a single "paths options.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 2]
```diff
@@ -180,6 +180,13 @@
         <xsd:attribute name="cache" type="xsd:string" />
         <xsd:attribute name="enable-annotations" type="xsd:boolean" />
         <xsd:attribute name="static-method" type="xsd:boolean" />
+        <xsd:attribute name="mapping" type="validation_mapping" />
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: `mapping` is no attribute (it must be an element)

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 3]
```diff
@@ -180,6 +180,13 @@
...
+    </xsd:complexType>
+
+    <xsd:complexType name="validation_mapping">
+        <xsd:sequence>
+            <xsd:element name="path" type="xsd:string" minOccurs="0" maxOccurs="unbounded" />
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: What about using `1` as the value for `minOccurs` (if you do not want to specify any paths you should omit the `mapping` element entirely imo)?

**Comment by xabbuh** [comment]
Status: Needs work

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd (hunk 3 above)
Comment (davewwww) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: i've never worked with xsd files. maybe you can give me a good solution for that.

Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: It should look something like this:

```xml
<xsd:complexType name="validation">
    <xsd:choice minOccurs="0" maxOccurs="unbounded">
        <xsd:element name="static-method" type="xsd:string" />
        <xsd:element name="mapping" type="validation_mapping" />
    </xsd:choice>

    <xsd:attribute name="enabled" type="xsd:boolean" />
    <xsd:attribute name="cache" type="xsd:string" />
    <xsd:attribute name="enable-annotations" type="xsd:boolean" />
    <xsd:attribute name="static-method" type="xsd:boolean" />
</xsd:complexType>

<xsd:complexType name="validation_mapping">
    <xsd:sequence>
        <xsd:element name="path" type="xsd:string" minOccurs="1" maxOccurs="unbounded" />
    </xsd:sequence>
</xsd:complexType>
```

**Comment by davewwww** [comment]
@xabbuh can I do anything else?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 4]
```diff
@@ -797,44 +800,76 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
...
+     * @param array $files  A array reference of validation files
+     */
+    private function getValidatorMappingFilesFromConfig(array $config, array &$files)
+    {
+        foreach ($config['mapping']['paths'] as $path) {
+            if (is_dir($path)) {
+                $this->getValidatorMappingFilesFromDir($path, $files);
+            } elseif (is_file($path)) {
+                if (preg_match('/\.(xml|ya?ml)$/', $path, $matches)) {
+                    $files[$matches[0]][] = realpath($path);
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `$matches[0]` will contain the string `yaml` when the file extension is `.yaml` which would now be ignored.

Comment (davewwww) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: you are right. i added a .yaml file extension support and added additional tests.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 5]
```diff
@@ -765,13 +765,16 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
...
+        $this->getValidatorMappingFilesFromConfig($config, $files);
+
+        if (!empty($files['xml'])) {
+            $validatorBuilder->addMethodCall('addXmlMappings', array($files['xml']));
         }
 
-        if (count($yamlMappings) > 0) {
-            $validatorBuilder->addMethodCall('addYamlMappings', array($yamlMappings));
+        if (!empty($files['yml']) || !empty($files['yaml'])) {
+            $validatorBuilder->addMethodCall('addYamlMappings', array(array_merge($files['yml'], $files['yaml'])));
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: What about calling the method twice instead to avoid the call to `array_merge()`?

Comment (davewwww) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: The idea was that there is only one call in the container.
But if you prefer the two calls, i will change it, no problem.

**Comment by lunetics** [comment]
This sounds like a good solution when not using bundles at all. Anything against merging that?

**Comment by davewwww** [comment]
@xabbuh: I would like to move on here, what can I do next?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 6]
```diff
@@ -797,44 +800,76 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
...
             }
         }
+    }
 
-        return $files;
+    /**
+     * Search for validation files in directory.
+     *
+     * @param string $dir   The directory that contains the validation files
+     * @param array  $files A array reference of validation files
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `An array`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 7]
```diff
@@ -797,44 +800,76 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
         }
     }
 
-    private function getValidatorMappingFiles(ContainerBuilder $container)
+    /**
+     * Get all validation files.
+     *
+     * @param ContainerBuilder $container A ContainerBuilder instance
+     * @param array            $files     A array reference of validation files
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `An array`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 8]
```diff
@@ -797,44 +800,76 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
...
+        foreach (Finder::create()->files()->in($dir)->name('/\.(xml|ya?ml)$/') as $file) {
+            $files[$file->getExtension()][] = $file->getRealpath();
+        }
+    }
+
+    /**
+     * Get all validation files from config.
+     *
+     * @param array $config A validation configuration array
+     * @param array $files  A array reference of validation files
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `an array`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 9]
```diff
@@ -797,44 +800,76 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
...
-
+                $this->getValidatorMappingFilesFromDir($dir, $files);
                 $container->addResource(new DirectoryResource($dir));
             }
         }
+    }
 
-        return $files;
+    /**
+     * Search for validation files in directory.
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Searches

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 10]
```diff
@@ -797,44 +800,76 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
...
+     */
+    private function getValidatorMappingFilesFromDir($dir, array &$files)
+    {
+        foreach (Finder::create()->files()->in($dir)->name('/\.(xml|ya?ml)$/') as $file) {
+            $files[$file->getExtension()][] = $file->getRealpath();
+        }
+    }
+
+    /**
+     * Get all validation files from config.
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Gets

**Comment by fabpot** [comment]
Can you add a note about this new feature in the CHANGELOG?

**Comment by davewwww** [comment]
of course, on which version?

File: CHANGELOG-3.0.md [hunk 11]
```diff
@@ -37,6 +37,7 @@ To get the diff between two versions, go to https://github.com/symfony/symfony/c
  * bug #18761  [Form] Modified iterator_to_array's 2nd parameter to false in ViolationMapper (issei-m)
  * bug #18745 [MonologBridge] Uninstallable together with symfony/http-kernel in 3.0.6 (ymc-dabe)
  * bug #18737 [Debug] Fix fatal error handlers on PHP 7 (nicolas-grekas)
+ * feature #19086 [FrameworkBundle] added configurable paths for validation files
```
Comment (fabpot) [inline CHANGELOG-3.0.md]: This should be removed. I was talking about the CHANGELOG in FrameworkBundle. This one is auto-generated when I make a release. Sorry about the confusion.

Comment (davewwww) [inline CHANGELOG-3.0.md]: ok, i've changed it. I'm not familiar with this process.

**Comment by fabpot** [comment]
Can you rebase on current master?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 12]
```diff
@@ -797,44 +800,76 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
...
+    /**
+     * Searches for validation files in directory.
+     *
+     * @param string $dir   The directory that contains the validation files
+     * @param array  $files An array reference of validation files
+     */
+    private function getValidatorMappingFilesFromDir($dir, array &$files)
+    {
+        foreach (Finder::create()->files()->in($dir)->name('/\.(xml|ya?ml)$/') as $file) {
+            $files[$file->getExtension()][] = $file->getRealpath();
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Can't you merge directly `yml` and `yaml` files here to avoid the `array_merge` later?

```php
$extension = $file->getExtension();
$files['yml' === $extension ? 'yaml' : $extension][] = $file->getRealpath();
```

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 13]
```diff
@@ -913,7 +948,7 @@ private function registerSerializerConfiguration(array $config, ContainerBuilder
             // Run after serializer.normalizer.object
             $definition = $container->register('serializer.normalizer.data_uri', DataUriNormalizer::class);
             $definition->setPublic(false);
-            $definition->addTag('serializer.normalizer', ['priority' => -920]);
+            $definition->addTag('serializer.normalizer', array('priority' => -920));
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: good catch (my bad), it should probably be moved to another PR targeting the 3.1 branch.

**Comment by dunglas** [comment]
👍 (when rebased). I left some (optional) minor comments.

**Comment by davewwww** [comment]
rebased and added the suggestion from dunglas

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 14]
```diff
@@ -34,7 +35,7 @@ CHANGELOG
  * Added `Symfony\Component\HttpHernel\DataCollector\RequestDataCollector::onKernelResponse()`
  * Added `Symfony\Bundle\FrameworkBundle\DataCollector\RequestDataCollector`
  * Deprecated service `serializer.mapping.cache.apc` (use `serializer.mapping.cache.doctrine.apc` instead)
-
+
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: should be reverted

**Comment by xabbuh** [comment]
👍 

Status: Reviewed

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 15]
```diff
@@ -979,41 +982,76 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
         }
     }
 
-    private function getValidatorMappingFiles(ContainerBuilder $container)
+    /**
+     * Get all validation files.
```
Comment (HeahDude) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Gets

**Review by HeahDude — APPROVED** [review]
👍

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 15 above)
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I think the whole phpdoc block can be removed. It does not add that much value (and is for a private method anyway).

Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Same for the other ones.

**Comment by fabpot** [comment]
Tests do not pass on Windows

**Comment by davewwww** [comment]
Yes, because
TestBundle\Resources\config\validation.xml
is not:
TestBundle/Resources/config/validation.xml

```php
$this->assertStringEndsWith('TestBundle/Resources/config/validation.xml', $xmlMappings[1]);
```
appears the error because of my changes?

**Comment by fabpot** [comment]
@davewwww Apparently, yes

**Comment by davewwww** [comment]
The problem is solved when I remove the realpath()

```php
if (is_file($file = $dirname.'/Resources/config/validation.xml')) {
   $files['xml'][] = realpath($file);
 }
```

**Comment by fabpot** [comment]
@davewwww  This now looks good to me. But before I can merge, can you squash your commit or at least rebase on current master (to get rid of the merge commit which prevents me to merge this PR cleanly).

Also, can you submit a PR to update the Symfony docs? Between, can you also change the description of this PR to reflect the latest version of this PR?

Thanks

**Comment by davewwww** [comment]
spuashed, pr updated and doc pr submitted

**Comment by fabpot** [comment]
Thank you @davewwww.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 16]
```diff
@@ -987,41 +990,58 @@ private function registerValidationConfiguration(array $config, ContainerBuilder
...
+        }
+    }
 
-        return $files;
+    private function getValidatorMappingFilesFromConfig(array $config, array &$files)
+    {
+        foreach ($config['mapping']['paths'] as $path) {
+            if (is_dir($path)) {
+                $this->getValidatorMappingFilesFromDir($path, $files);
+            } elseif (is_file($path)) {
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: this must use the ContainerBuilder API to register resources to invalidate the cache when files are changing in debug mode

Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: indeed, see #21927
